Source/WebCore/animation/AnimationTimelinesController.cpp

```diff
@@ -310,42 +311,79 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    // A named scroll progress timeline or view progress timeline is referenceable by:
+    // 1. the name-declaring element itself
+    // 2. that element’s descendants
+    // If multiple elements have declared the same timeline name, the matching timeline is the one declared on the nearest element in tree order.
+    // In case of a name conflict on the same element, names declared later in the naming property (scroll-timeline-name, view-timeline-name) take
+    // precedence, and scroll progress timelines take precedence over view progress timelines.
+
+    Vector<Ref<AnimationTimeline>> elementAncestors;
+    for (auto& entry : entries) {
+        if (entry->originatingElement() ==  &matchElement || matchElement.isDescendantOf(entry->originatingElement()))
+            elementAncestors.append(entry);
+    }
+
+    if (elementAncestors.isEmpty())
+        return nullptr;
+    if (elementAncestors.size() == 1)
+        return RefPtr { elementAncestors.first().ptr() };
+    return determineTreeOrder(elementAncestors, matchElement);
 }
 
-ScrollTimeline* AnimationTimelinesController::scrollTimelineForName(const AtomString& name) const
+Vector<Ref<AnimationTimeline>>& AnimationTimelinesController::timelinesForName(const AtomString& name)
 {
-    return m_nameToScrollTimelineMap.get(name);
+    return m_nameToTimelineMap.ensure(name, [] {
+        return Vector<Ref<AnimationTimeline>> { };
+    }).iterator->value;
+}
+
+void AnimationTimelinesController::registerNamedScrollTimeline(const AtomString& name, Element& source, ScrollAxis axis)
+{
+    auto& timelines = timelinesForName(name);
+
+    auto existingTimelineIndex = timelines.findIf([&](auto& timeline) {
+        return is<ScrollTimeline>(timeline) && timeline->originatingElement() == &source;
+    });
+
+    if (existingTimelineIndex != notFound) {
+        auto existingScrollTimeline = downcast<ScrollTimeline>(timelines[existingTimelineIndex]);
+        existingScrollTimeline->setAxis(axis);
+    } else {
+        auto newScrollTimeline = ScrollTimeline::create(name, axis);
+        newScrollTimeline->setSource(&source);
+        timelines.append(WTFMove(newScrollTimeline));
+    }
 }
 
 void AnimationTimelinesController::registerNamedViewTimeline(const AtomString& name, Element& subject, ScrollAxis axis, ViewTimelineInsets&& insets)
 {
-    auto& timelines = m_nameToViewTimelinesMap.ensure(name, [] {
-        return Vector<Ref<ViewTimeline>> { };
-    }).iterator->value;
+    auto& timelines = timelinesForName(name);
 
     auto existingTimelineIndex = timelines.findIf([&](auto& timeline) {
-        return timeline->subject() == &subject;
+        return is<ViewTimeline>(timeline) && timeline->originatingElement() == &subject;
     });
 
     if (existingTimelineIndex != notFound) {
-        auto& existingViewTimeline = timelines[existingTimelineIndex];
+        auto existingViewTimeline = downcast<ViewTimeline>(timelines[existingTimelineIndex]);
```

Comment:
I believe this creates a second `Ref`, please use a reference like the code did before.

Reply:
Fixed.